pkg/solana/chainreader/event_read_binding.go

```diff
@@ -115,6 +158,56 @@ func (b *eventReadBinding) QueryKey(
 	return sequences, nil
 }
 
+func (b *eventReadBinding) normalizeParams(value any, itemType string) (any, error) {
+	offChain, err := b.codec.CreateType(itemType, true)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to create type: %w", types.ErrInvalidType, err)
+	}
+
+	// params can be a singular primitive value, a map of values, or a struct
+	// in the case that the input params are presented as a map of values, apply the values to the off-chain type
+	// with solana hooks
+	if err = codec.MapstructureDecode(value, offChain); err != nil {
+		return nil, fmt.Errorf("%w: failed to decode offChain value: %s", types.ErrInternal, err.Error())
+	}
+
+	return offChain, nil
+}
+
+func (b *eventReadBinding) extractFilterSubkeys(offChainValue any) (query.Expression, error) {
+	var expressions []query.Expression
+
+	for offChainKey, idx := range b.indexedSubKeys {
+		itemType := codec.WrapItemType(true, b.namespace, b.genericName+"."+offChainKey)
+
+		onChainValue, err := b.modifier.TransformToOnChain(offChainValue, itemType)
+		if err != nil {
+			return query.Expression{}, fmt.Errorf("%w: failed to apply on-chain transformation for key %s", types.ErrInternal, offChainKey)
+		}
+
+		// check that onChainValue is not zero value for type
+		if reflect.ValueOf(onChainValue).IsZero() {
+			continue
+		}
```

Comment:
Why do we treat zero as special, don't we still need to convert the type from onChain to offChain?

Reply:
Because a zero value could make it's way to the filter without the caller having set it to be included. It's the consequence of using struct types for filtering. A very common Go ORM, `gorm`, handles filtering this way too. Zero values are ignored when passing structs.

Reply:
I'm also fairly certain this is how the EVM implementation works too.

Reply:
Ah, ok if they're always converted to pointers that makes sense.  I just wanted to make sure that it's possible for someone to search for, say, subKeyValue1 = 0 and get back the right results